napari/_vispy/canvas.py

```diff
@@ -96,3 +98,88 @@ def _process_mouse_event(self, event):
         if event.type == 'mouse_wheel' and len(event.modifiers) > 0:
             return
         super()._process_mouse_event(event)
+
+
+class FramerateMonitor:
+    """Tracks and filters the framerate emitted from the canvas measure_fps() callback."""
+
+    def __init__(
+        self,
+        fps_window: float = 0.5,
+        stale_threshold: float = 0.6,
+        debounce_threshold: int = 2,
+    ):
+
+        self.events = EmitterGroup(source=self, fps=Event)
+        self._fps_window = fps_window
+        self._debounce_counter = 0
+        self._debounce_threshold = debounce_threshold
+        self._last_update = time.time()
+        self._stale_threshold = stale_threshold
+
+        self._fps = 0
+        self._measuring = False
+
+        self._last_measurement_valid = False
+
+    @property
+    def fps(self) -> float:
+        """The most recently measure framerate in frames per second."""
+        return self._fps
+
+    @property
+    def valid(self) -> bool:
+        """Flag set to True if the current fps measurement is valid."""
+        return self._last_measurement_valid and not self._fps_stale()
+
+    def _fps_stale(self):
+        """Check if the too much time has elapsed since the last fps update.
+
+        Returns
+        -------
+        fps_stale : bool
+            Flag set to True if the time since the last update is greater
+            than the _stale_threshold
+        """
+        return (time.time() - self._last_update) > self._stale_threshold
+
+    def update_fps(self, fps: float):
+        """Update with the most recently measured framerate.
+
+        This only stores the new framerate if the last draw was within
+        the specified stale_threshold and the debounce condition has
+        been met.
+
+        If the framerate update is valid, the fps event is emitted.
+
+        This is generally connected to the canvas.measure_fps() callback.
+
+        Parameters
+        ----------
+        fps : float
+            The newly measured framerate in frames per second.
+        """
+        if not self._fps_stale():
+            # do nothing if the last fps measurement is still valid
+            return
+        elif self._measuring is False:
```

Comment:
Just a small nit-pick.

```suggestion
        elif not self._measuring:
```